pkg/lib/infra/sms/provider_resolver_test.go

```diff
@@ -0,0 +1,126 @@
+package sms
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/kelseyhightower/envconfig"
+	. "github.com/smartystreets/goconvey/convey"
+	goyaml "gopkg.in/yaml.v2"
+	"sigs.k8s.io/yaml"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+)
+
+// nolint: gocognit
+func TestProviderResolver(t *testing.T) {
+	Convey("resolve provider", t, func() {
+		f, err := os.Open("testdata/provider_resolver_tests.yaml")
+		if err != nil {
+			panic(err)
+		}
+		defer f.Close()
+
+		type AuthgesrSecretsYAML struct {
+			Messaging *config.MessagingConfig         `json:"messaging"`
+			Nexmo     *config.NexmoCredentials        `json:"nexmo"`
+			Twilio    *config.TwilioCredentials       `json:"twilio"`
+			Custom    *config.CustomSMSProviderConfig `json:"custom"`
+		}
+
+		type EnvConfig struct {
+			SMSGateway *config.SMSGatewayConfig `envconfig:"SMS_GATEWAY"`
+		}
+
+		type TestCase struct {
+			Name                 string      `yaml:"name"`
+			Result               string      `yaml:"result"`
+			AuthgearSecretsYAML  interface{} `yaml:"authgear.secrets.yaml"`
```

Comment:
The testcase should take `authgear.yaml`, `authgear.secrets.yaml`, and `environment_variables`.